### PR TITLE
add logic for openshift webconsole operator

### PR DESCRIPTION
**Pull request by deads2k**
Adds a simple operator that creates a webconsole deployment.  It manages conflicts with the workload controller by actively stomping it when the generation doesn't match.  Errors appear on the status of the subresource

### CODE REVIEW

File: pkg/cmd/openshift-operators/util/resourceapply/apps.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+	if *modified {
+		actual, err := client.Deployments(required.Namespace).Update(existing)
+		return actual, true, err
+	}
+	if existing.ObjectMeta.Generation == expectedGeneration {
+		return existing, false, nil
+	}
+	existing.Spec = *required.Spec.DeepCopy()
+
+	actual, err := client.Deployments(required.Namespace).Update(existing)
```
Comment (mfojtik) [inline pkg/cmd/openshift-operators/util/resourceapply/apps.go]: where do we handle conflicts? does the controller just re-apply on next sync?

File: pkg/cmd/openshift-operators/webconsole-operator/v310_00/deployment.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+kind: Deployment
+metadata:
+  namespace: openshift-web-console
+  name: webconsole
+  labels:
+    app: openshift-web-console
+    webconsole: "true"
+spec:
+  strategy:
+    type: Recreate
```
Comment (mfojtik) [inline pkg/cmd/openshift-operators/webconsole-operator/v310_00/deployment.go]: is there reason this is Recreate and not Rollout ?

Comment (soltysh) [reply pkg/cmd/openshift-operators/webconsole-operator/v310_00/deployment.go]: It was discussed some time ago, web-console switched to recreate some time ago, iirc.

File: pkg/cmd/openshift-operators/util/resourceapply/apps.go (hunk 1 above)
Comment (deads2k) [reply pkg/cmd/openshift-operators/util/resourceapply/apps.go]: > where do we handle conflicts? does the controller just re-apply on next sync?

Yes, and it is logged and put into the CRD status.

File: pkg/cmd/openshift-operators/util/resourceread/core.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+		panic(err)
+	}
+}
+
+func ReadConfigMapOrDie(objBytes []byte) *corev1.ConfigMap {
+	requiredObj, err := runtime.Decode(coreCodecs.UniversalDecoder(corev1.SchemeGroupVersion), []byte(objBytes))
+	if err != nil {
+		panic(err)
+	}
+	return requiredObj.(*corev1.ConfigMap)
```
[the remark this answers is not in the document]
Comment (deads2k) [reply pkg/cmd/openshift-operators/util/resourceread/core.go]: > can we also panic when requiredObj != corev1.ConfigMap ?

Yeah.  it's a coding error.

**Comment by deads2k** [comment]
@mfojtik @jpeeler any known problems with service catalog?  Doesn't immediately look related to my changes.

**Comment by jpeeler** [comment]
@deads2k There is, this PR needs to be merged: https://github.com/openshift/origin/pull/19568. I really expected for it to land overnight. ..
Btw, what link are you looking at for failing catalog tests? Reporting has been disabled temporarily until the above PR merges.

**Comment by deads2k** [comment]
> @deads2k There is, this PR needs to be merged: #19568. I really expected for it to land overnight. ..
> Btw, what link are you looking at for failing catalog tests? Reporting has been disabled temporarily until the above PR merges.

Cluster up tests when you modify cluster up indicate it here: https://openshift-gce-devel.appspot.com/build/origin-ci-test/pr-logs/pull/19498/test_pull_request_origin_extended_clusterup/1021/

**Comment by jpeeler** [comment]
/retest
I saw the same error with catalog, then pulled/updated everything and now it seems to have gone away.

**Comment by jpeeler** [comment]
I can't figure out why catalog is failing to start in the extended clusterup test. Other PRs seem to be successfully passing.

**Comment by mfojtik** [comment]
@jpeeler the failure seems to be `Failed to install "openshift-web-console-operator": timed out waiting for the condition` (looking at that now), the service catalog messages are just noise

File: pkg/oc/bootstrap/clusteradd/components/web-console-operator/web_console_operator.go [hunk 4]
```diff
@@ -48,7 +53,21 @@ func (c *WebConsoleOperatorComponentOptions) Install(dockerClient dockerhelper.I
 		RBACTemplate:    bootstrap.MustAsset("install/openshift-web-console-operator/install-rbac.yaml"),
 		InstallTemplate: bootstrap.MustAsset("install/openshift-web-console-operator/install.yaml"),
 
-		// TODO wait until the webconsole is up
+		// wait until the webconsole to an available endpoint
+		WaitCondition: func() (bool, error) {
+			glog.V(2).Infof("polling for service catalog api server endpoint availability")
```
Comment (mfojtik) [inline pkg/oc/bootstrap/clusteradd/components/web-console-operator/web_console_operator.go]: polling for webconsole availability?

**Comment by mfojtik** [comment]
@deads2k I think this is failing because you use the Always as the OPENSHIFT_PULL_POLICY which means the `hypershift` does not have the `experimental` subcommand (at least that is the case for my locally when I will get the same timeout...). Will make a PR against your PR.

**Comment by deads2k** [comment]
ah, image layout changed

**Comment by mfojtik** [comment]
/lgtm

this is a good starting point

**Pull request merged**